src/java/org/apache/cassandra/index/IndexStatusManager.java

```diff
@@ -84,26 +85,39 @@ private IndexStatusManager()
      */
     public <E extends Endpoints<E>> E filterForQuery(E liveEndpoints, Keyspace keyspace, Index.QueryPlan indexQueryPlan, ConsistencyLevel level)
```

Comment:
The next call after `filterForQuery` is to filter for EACH_QUOUM in `ReplicaPlans#contactForRead`. If 1st DC has 3 succeeded replicas and 2nd DC has 3 unknown repliacs. `filterForQuery` will return 3 succeeded + 1 unknown and will fail the consistency requirement for `EACH_QUOUM`. Not sure if we care about EACH_QUOUM for indexes.

The safest approach here is to sort replicas by query status. How about following?

```
        Set<Replica> queryableNonSucceeded = new HashSet<>(liveEndpoints.size());
        E queryableEndpoints = liveEndpoints.filter(replica -> {

            boolean allBuilt = true;
            for (Index index : indexQueryPlan.getIndexes())
            {
                Index.Status status = getIndexStatus(replica.endpoint(), keyspace.getName(), index.getIndexMetadata().name);
                if (!index.isQueryable(status))
                    return false;

                if (status != Index.Status.BUILD_SUCCEEDED)
                    allBuilt = false;
            }

            if (!allBuilt)
                queryableNonSucceeded.add(replica);

            return true;
        });

        // deprioritize replica with queryable but non-succeeded state
        if (!queryableNonSucceeded.isEmpty())
            queryableEndpoints = queryableEndpoints.sorted(Comparator.comparingInt(r -> queryableNonSucceeded.contains(r) ? 1 : -1));
```

Reply:
`contactForRead()` just generates a sub-list from left to right, so that should work. In most cases, the sort shouldn't even happen, as `UNKNOWN` states should be transient/rare. Since the number of nodes in this state should very small, and we may have a large cluster, I'd also make the initial size of `queryableNonSucceeded` something reasonably small, like 4 (w/ a comment), or even use the default (16). With the `ReplicaCollection.Builder` it made sense to do the opposite, but that, in retrospect, would not have been very nice either in a large cluster with most things built/succeeded (aside from avoiding resizing).

Reply:
For partition-restricted index queries, the `liveEndpoints` might be small, of course, but I'd still size `queryableNonSucceeded` explicitly at something like 4.

Reply:
Addressed in the latest commit. Thanks for the review!